### PR TITLE
fix updating assigned identity when azure identity updated

### PR DESCRIPTION
**Pull request by aramase**
**Reason for Change**:

- Adds code path to update azure assigned identity when azure identity is updated in place
- Ensures old identity is deleted and new updated identity is assigned
- Adds e2e test to test update path

**Issue Fixed**:

fixes https://github.com/Azure/aad-pod-identity/issues/510, fixes https://github.com/Azure/aad-pod-identity/issues/504

**Notes for Reviewers**:

### CODE REVIEW

File: pkg/cloudprovider/cloudprovider.go [hunk 1]
```diff
@@ -178,18 +178,19 @@ func (c *Client) UpdateUserMSI(addUserAssignedMSIIDs, removeUserAssignedMSIIDs [
...
-	// add new ids to the list
-	for _, userAssignedMSIID := range addUserAssignedMSIIDs {
-		ids[userAssignedMSIID] = true
-	}
 	// remove msi ids from the list
 	for _, userAssignedMSIID := range removeUserAssignedMSIIDs {
 		ids[userAssignedMSIID] = false
 	}
+	// add new ids to the list
+	for _, userAssignedMSIID := range addUserAssignedMSIIDs {
```
Comment (kkmsft) [inline pkg/cloudprovider/cloudprovider.go:186]: Is there. change here ?

Comment (aramase) [reply pkg/cloudprovider/cloudprovider.go:186]: Yeah, the remove needs to be done before assign. The order is required to ensure we don't delete an identity that needs to be assigned.

Comment (kkmsft) [reply pkg/cloudprovider/cloudprovider.go:186]: This is part of #564 now - right ?

Comment (aramase) [reply pkg/cloudprovider/cloudprovider.go]: Yeah, rebased the PR now on the master.

File: pkg/mic/mic.go [hunk 2]
```diff
@@ -428,24 +429,25 @@ func (c *Client) Sync(exit <-chan struct{}) {
...
 		}
-		klog.V(5).Infof("del: %v, add: %v", deleteList, addList)
+		beforeUpdateList, afterUpdateList := c.getAzureAssignedIdentitiesToUpdate(addList, deleteList)
+		klog.V(5).Infof("del: %v, add: %v, update: %v", deleteList, addList, afterUpdateList)
 
 		// the node map is used to track assigned ids to create/delete, identities to assign/remove
 		// for each node or vmss
 		nodeMap := make(map[string]trackUserAssignedMSIIds)
 
 		// seperate the add and delete list per node
```
Comment (chewong) [inline pkg/mic/mic.go:439]: - s/seperate/separate
- separate the add, delete and update list per node

File: pkg/crd/crd.go [hunk 3]
```diff
@@ -519,6 +520,32 @@ func (c *Client) CreateAssignedIdentity(assignedIdentity *aadpodid.AzureAssigned
...
+	klog.Infof("Got assigned id %s/%s to update", assignedIdentity.Namespace, assignedIdentity.Name)
+	begin := time.Now()
+
+	defer func() {
+		if err != nil {
+			c.reporter.ReportKubernetesAPIOperationError(metrics.AssignedIdentityAdditionOperationName)
+			return
+		}
+		c.reporter.Report(
+			metrics.AssignedIdentityAdditionCountM.M(1),
```
[the remark this answers is not in the document]
Comment (aramase) [reply pkg/crd/crd.go:533]: Makes sense! My initial assumption was to use this as a tracker for # of identities, but this metric signifies the # of assigned id operations. Will update it!

File: pkg/metrics/metrics.go [hunk 4]
```diff
@@ -46,6 +48,8 @@ const (
 	AssignedIdentityDeletionOperationName = "assigned_identity_deletion"
 	// AssignedIdentityAdditionOperationName ...
 	AssignedIdentityAdditionOperationName = "assigned_identity_addition"
+	// AssignedIdentityUpdateOperationName ...
```
Comment (chewong) [inline pkg/metrics/metrics.go:51]: Let's open another PR to clean up these `...` comments as I noticed there are a lot in the codebase.

Comment (aramase) [reply pkg/metrics/metrics.go:51]: Opened an issue to track this - https://github.com/Azure/aad-pod-identity/issues/571. We can use this as a parent issue and add sub items as we make the changes.

**Pull request merged**